src/java.base/unix/classes/sun/nio/fs/UnixFileSystem.java

```diff
@@ -116,6 +96,31 @@ byte[] defaultDirectory() {
     }
 
     boolean needToResolveAgainstDefaultDirectory() {
+        // OK if two or more threads initialize this instance variable
+        if (needToResolveAgainstDefaultDirectory == null) {
+            // if process-wide chdir is allowed or default directory is not the
+            // process working directory then paths must be resolved against the
+            // default directory.
+            String propValue = GetPropertyAction
+                .privilegedGetProperty("sun.nio.fs.chdirAllowed", "false");
+            boolean chdirAllowed = Boolean.parseBoolean(propValue);
+
+            if (chdirAllowed) {
+                this.needToResolveAgainstDefaultDirectory = true;
+            } else {
+                byte[] cwd = UnixNativeDispatcher.getcwd();
+                boolean defaultIsCwd = (cwd.length == defaultDirectory.length);
+                if (defaultIsCwd) {
+                    for (int i=0; i<cwd.length; i++) {
+                        if (cwd[i] != defaultDirectory[i]) {
+                            defaultIsCwd = false;
+                            break;
+                        }
+                    }
+                }
+                this.needToResolveAgainstDefaultDirectory = !defaultIsCwd;
+            }
+        }
```

Comment:
Making needToResolveAgainstDefaultDirectory a tri-state is a bit ugly. If we really need to defer this then it might be better to just move the field to Holder class, that way this method would return holder.needToResolveAgainstDefaultDirectory.

Reply:
The use of `needToResolveAgainstDefaultDirectory` is removed in 267f9a699410cc39fefeedd9623eae756f4e1fcd.